app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/DeviceListFragment.java

```diff
@@ -0,0 +1,357 @@
+      showToast(localFileTransferActivity, R.string.preparing_to_receive, Toast.LENGTH_SHORT);
+
+    } else if(groupInfo.groupFormed) {
+      {
+        Log.d(LocalFileTransferActivity.TAG, "Starting file transfer");
+
+        showToast(localFileTransferActivity, R.string.starting_transfer, Toast.LENGTH_SHORT);
+
+        if(groupInfo.isGroupOwner)  fileReceiverDeviceAddress = selectedPeerDeviceInetAddress;
+        else                        fileReceiverDeviceAddress = groupInfo.groupOwnerAddress;
+
+        // Hack for allowing slower receiver devices to setup server before sender device requests to connect
+        showToast(localFileTransferActivity, R.string.preparing_files, Toast.LENGTH_LONG);
+        for(int i = 0; i < 20000000; i++);
+
+        for(int i = 0; i < totalFilesForTransfer; i++) {
+          new SenderDeviceAsyncTask(this, transferProgressFragment, i).execute(fileUriList.get(i));
+        }
+      }
+    }
+  }
+
+  /* Helper methods */
+
+  public WifiP2pDevice getUserDevice() {
+    return userDevice;
+  }
+
+  public boolean isFileSender() {
+    return fileSender;
+  }
+
+  public int getTotalFilesForTransfer() {
+    return totalFilesForTransfer;
+  }
+
+  public void setTotalFilesForTransfer(int totalFilesForTransfer) {
+    this.totalFilesForTransfer = totalFilesForTransfer;
+  }
+
+  public ArrayList<FileItem> getFileItems() {
+    return filesToSend;
+  }
+
+  public void setFileItems(ArrayList<FileItem> fileItems) {
+    this.filesToSend = fileItems;
+  }
+
+  public ArrayList<Uri> getFileUriList() {
+    return fileUriList;
+  }
+
+  public void incrementTotalFilesSent() {
+    this.filesSent++;
+  }
+
+  public boolean allFilesSent() {
+    return (filesSent == totalFilesForTransfer);
+  }
+
+  public void setSharedPreferenceUtil(SharedPreferenceUtil sharedPreferenceUtil) {
+    this.sharedPreferenceUtil = sharedPreferenceUtil;
+  }
+
+  public String getZimStorageRootPath() {
+    return (sharedPreferenceUtil.getPrefStorage() + "/Kiwix/");
+  }
+
+  public InetAddress getFileReceiverDeviceAddress() {
+    return fileReceiverDeviceAddress;
+  }
+
+
+  public static String getFileName(Uri fileUri) {
+    String fileUriString = fileUri.toString();
+    // Returns text after location of last slash in the file path
+    return fileUriString.substring(fileUriString.lastIndexOf('/')+1);
+  }
+
+  public static void copyToOutputStream(InputStream inputStream, OutputStream outputStream) throws IOException {
+    byte[] bufferForBytes = new byte[1024];
+    int bytesRead;
+
+    Log.d(TAG, "Copying to OutputStream...");
+    while ((bytesRead = inputStream.read(bufferForBytes)) != -1) {
+      outputStream.write(bufferForBytes, 0, bytesRead);
+    }
+
+    outputStream.close();
+    inputStream.close();
+    Log.d(LocalFileTransferActivity.TAG, "Both streams closed");
+  }
+
+
+  private class WifiPeerListAdapter extends ArrayAdapter<WifiP2pDevice> {
```

Comment:
I would prefer this to be a separate class.
Ideally this is a recylerview too but I won't force the issue

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/1e74176c852cd5ecfc400a5cfd3ca7d5aaf3cb90)
This adapter is for the ListFragment displaying list of peer devices, and since I was using it for the first time I didn't want to digress much from the tutorial to keep myself from missing the deadline
Should I convert it to use recyclerview?

Reply:
Recyclerviews are just better, I view ListView as long deprecated and we do have a ticket to replace all listviews? #613 that is just the library fragment which I have already done so I closed it.
I would say use the Recyclerview, it is a better learning experience for yourself as well. That said I don't know what deadlines are in place.

Reply:
In my proposal I had kept notekeeper and localfiletransfer within the first phase, which ended on 28 June (shining example of my pro-level estimation skills)
I am almost done with the smaller review changes, so I think I'll do the conversion with the major changes that seem to be required from your comment [here](https://github.com/kiwix/kiwix-android/pull/1276/#discussion_r301945332)